### PR TITLE
gh-102450: Add ISO-8601 alternative for midnight to `fromisoformat()` calls.

### LINKED ISSUE #102450: datetime library doesn't support valid ISO-8601 alternative for midnight
# Bug report

According to [ISO-8601](https://en.wikipedia.org/wiki/ISO_8601), a time of `24:00` on a given date is a valid alternative to `00:00` of the following date, however Python does not support this, raising the following error when attempted: `ValueError: hour must be in 0..23`.

This bug can be seen from multiple scenarios, specifically anything that internally calls the [`_check_time_fields` function](https://github.com/python/cpython/blob/main/Lib/datetime.py#L531-L546), such as the following:
```python
>>> import datetime
>>> datetime.datetime(2022, 1, 2, 24, 0, 0)  # should be equivalent to 2022-01-03 00:00:00
Traceback (most recent call last):
  File "<pyshell#2>", line 1, in <module>
    datetime.datetime(2022, 3, 4, 24, 0, 0)
ValueError: hour must be in 0..23
```
The fix for this is relatively simple: have an explicit check within [`_check_time_fields`](https://github.com/python/cpython/blob/main/Lib/datetime.py#L531-L546) for the scenario where `hour == 24 and minute == 0 and second == 0 and microsecond == 0`, or more concisely `hour == 24 and not any((minute, second, microsecond))`, and in this scenario increase the day by one (adjusting the week/month/year as necessary) and set the hour to 0.

I imagine the [`check_time_args` C function](https://github.com/python/cpython/blob/main/Modules/_datetimemodule.c#L467-L499) would also have to be updated.

# Your environment

- CPython versions tested on: 3.9.12, 3.10.7, 3.11.2 (presumably applies to all)
- Operating system and architecture: MacOS Ventura arm64 (presumably applies to all)

### Linked PRs
* gh-105856

### LINKED ISSUE #124257: Invalid  datetime.time.fromisoformat
# Bug report

### Bug description:

According to the standard ISO 8601 `Thus the value ‘24:00:00’ is chosen to represent the final
instant of the day in order to distinguish the expression of this instant from an expression of a time scale
unit.` 

```python
>>> datetime.time.fromisoformat("24:00")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: hour must be in 0..23

```
[Link to standard pdf](https://cdn.standards.iteh.ai/samples/81801/f527872a9fe34281ae3a4af8e730f3f8/ISO-8601-1-2019-Amd-1-2022.pdf)

### CPython versions tested on:

3.10, 3.12

### Operating systems tested on:

Linux, macOS

### PR DESCRIPTION
**Pull request by TizzySaurus**
Closes #102450.
Closes #124257.

I've updated `_datetimemodule.c` and `_pydatetime.py` such that calls to Python's `datetime.time.fromisoformat` and `datetime.datetime.fromisoformat` both allow providing the `24:00` ISO-8601 alternative to midnight, but the class constructors themselves do not (as per [this comment from @pganssle](https://github.com/python/cpython/issues/102450#issuecomment-1547909134)).

Screenshots of the C code working:
datetime.time.fromisoformat()
datetime.datetime.fromisoformat()

I've also done the same tests for the `_pydatetime` module and everything appears to be working.

Also updated the datetime tests accordingly, to ensure that `datetime.time.fromisoformat("24:00:00.000000")` is parsed as `datetime.time(0, 0, 0, 0)` etc. and the cases that should error (such as `datetime.time.fromisoformat("24:30")`) do indeed error.

* Issue: gh-102450

### CODE REVIEW

**Comment by wookie184** [comment]
> It seems I don't need to edit the Python code, but just the C?

The PEP that explains it was linked on the issue: https://github.com/python/cpython/issues/102450#issuecomment-1553081049

Basically, (I believe) you need to implement it in both places but only the C version is used by CPython. If you add tests they should run on both implementations so you can ensure you've done things correctly. You should also be able to `import _pydatetime` directly to test it manually.

**Comment by TizzySaurus** [comment]
> The PEP that explains it was linked on the issue: [#102450 (comment)](https://github.com/python/cpython/issues/102450#issuecomment-1553081049)

> Basically, (I believe) you need to implement it in both places but only the C version is used by CPython. If you add tests they should run on both implementations so you can ensure you've done things correctly. You should also be able to `import _pydatetime` directly to test it manually.

**Comment by TizzySaurus** [comment]
The tests that I added in 311f914 are actually passing, despite me not having implemented the feature into the python code yet. My understanding of [PEP399](https://peps.python.org/pep-0399/) was that the one file tested both the Python and C implementation, but it seems that `datetimetester.py` actually _only_ tests the C code? Do I have to write the tests for the Python implementation of the library somewhere else?

**Comment by TizzySaurus** [comment]
I've also just noticed that the `datetime` and `_pydatetime` modules give different errors when specifying an invalid day:
```python
>>> from datetime import datetime as c_datetime
>>> c_datetime.fromisoformat("2023-01-32T24:00:00")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: day is out of range for month
```
vs
```python
>>> from _pydatetime import datetime as py_datetime
>>> py_datetime.fromisoformat("2023-01-32T24:00:00")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/xxx/Documents/personal/cpython/Lib/_pydatetime.py", line 1906, in fromisoformat
    return cls(*(date_components + time_components))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xxx/Documents/personal/cpython/Lib/_pydatetime.py", line 1717, in __new__
    year, month, day = _check_date_fields(year, month, day)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xxx/Documents/personal/cpython/Lib/_pydatetime.py", line 545, in _check_date_fields
    raise ValueError('day must be in 1..%d' % dim, day)
ValueError: ('day must be in 1..31', 32)
```
Note how `_pydatetime`'s error message specifies the actual day range for that month. Is this something I should add to the C implementation whilst I'm here?

File: Misc/NEWS.d/next/Library/2023-06-16-14-52-00.gh-issue-102450.MfeR6A.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Add missing ISO-8601 24:00 alternative to midnight of next day to :meth:`datetime.datetime.fromisoformat()` and :meth:`datetime.time.fromisoformat()`.
```
Comment (pganssle) [inline Misc/NEWS.d/next/Library/2023-06-16-14-52-00.gh-issue-102450.MfeR6A.rst]: Do you want to give yourself credit here?

Patch by ...

**Review by pganssle — APPROVED** [review]
This looks great! Do you want to add yourself to the ACKS file and/or to the patch name, so you get credit? (This is optional)

Also can you rebase against main?

**Review by vstinner — APPROVED** [review]
LGTM. I like how invalid inputs (month 13, non-zero minute, etc.) are checked: it's smart (and keep the code short) :-) I also like tests which seem to cover all cases.

About the feature itself, I trust @pganssle's opinion to decide if we should implement it or not.

File: Misc/NEWS.d/next/Library/2023-06-16-14-52-00.gh-issue-102450.MfeR6A.rst (hunk 1 above)
Comment (TizzySaurus) [reply Misc/NEWS.d/next/Library/2023-06-16-14-52-00.gh-issue-102450.MfeR6A.rst]: Done in [1b65671](https://github.com/python/cpython/pull/105856/commits/1b65671e346d454915578c5e502891dc40a60a4d)!

**Comment by TizzySaurus** [comment]
> @pganssle Also can you rebase against main?

Think I've done this correctly, so this should hopefully be good to merge now 🙂

File: Lib/test/datetimetester.py [hunk 2]
```diff
@@ -3378,6 +3381,11 @@ def test_fromisoformat_fails_datetime(self):
             '2009-04-19T12:30:45.123456-05:00a',    # Extra text
             '2009-04-19T12:30:45.123-05:00a',       # Extra text
             '2009-04-19T12:30:45-05:00a',           # Extra text
+            '2009-04-19T24:00:00.000001',  # Has non-zero microseconds on 24:00
+            '2009-04-19T24:00:01.000000',  # Has non-zero seconds on 24:00
+            '2009-04-19T24:01:00.000000',  # Has non-zero minutes on 24:00
+            '2009-04-32T24:00:00.000000',  # Day is invalid before wrapping due to 24:00
+            '2009-13-01T24:00:00.000000',  # Month is invalid before wrapping due to 24:00
```
Comment (vstinner) [inline Lib/test/datetimetester.py:3388]: You may also test `'9999-12-31T24:00:00'` which fails with "year 10000 is out of range".

**Review by vstinner — COMMENTED** [review]
Oh nice, the code handles also well the maximum date:

```
$ ./python
>>> import datetime
>>> datetime.datetime.fromisoformat('9999-12-31T24:00:00')
ValueError: year 10000 is out of range
```

File: Lib/test/datetimetester.py (hunk 2 above)
Comment (TizzySaurus) [reply Lib/test/datetimetester.py:3388]: Done in 2ccb2a220c :+1:

**Pull request merged**